docs: progressive override (#2904)
apollographql/federation: docs/source/entities-advanced.mdx

The [`@override` directive](./federated-types/federated-directives#override) enables you to incrementally migrate between subgraphs with no downtime.

Follow these steps to use the `@override` directive:

1. If the `@override` directive isn't already imported, include it in your schema's `@link` imports:

    ```graphql {3} title="Billing subgraph"
    extend schema
      @link(url: "https://specs.apollo.dev/federation/v2.3",
            import: ["@key", "@shareable", "@override"])
    ```

2. Deploy a new version of the Billing subgraph that both defines and resolves the `Bill` fields you want to move:

    <CodeColumns>

      ```graphql title="Payments subgraph"
      type Bill @key(fields: "id") {
        id: ID!
        amount: Int!
        payment: Payment
      }

      type Payment {
        # ...
      }
      ```

      ```graphql {3} title="Billing subgraph"
      type Bill @key(fields: "id") {
        id: ID!
        amount: Int! @override(from: "Payments") #highlight-line
      }
      ```

    </CodeColumns>

      Applying the `@override` directive tells the router to resolve the `amount` field in the Billing subgraph instead of the Payments subgraph.

> 3. Update your router's supergraph schema to migrate to the updated Billing subgraph, either gradually or all at once.

Author: emphasize enterprise-only feature of progressive override

    * **To migrate gradually**, use the [progressive `@override` feature](/router/configuration):

>       1. Configure your router to enable progressive `@override`:

Author: This step not needed, instead describe that router auto-detects usages of progressive override (has `labels`)

          ```yaml title="router.yaml"
          progressive_override:
            enabled: true
          ```
      1. In the Billing subgraph, add a `labels` argument to the `@override` directive with the percentage of requests that you want to be resolved by the subgraph. For example:

        ```graphql title="Billing subgraph"
        type Bill @key(fields: "id") {
          id: ID!
          amount: Int! @override(from: "Payments", labels:"percent(15)") #highlight-line
        }
        ``` 

>         Setting `labels:"percent(15)"` means that 15 percent of the requests for `amount` are resolved by the Billing subgraph, while the remaining 85 percent are resolved by the Payments subgraph. 

Author: replace `labels` --> `label`

</CodeColumns>

The [`@override` directive](./federated-types/federated-directives#override) enables you to incrementally migrate between subgraphs with no downtime.

You can migrate between subgraphs gradually with progressive `@override` or all at once.

<ProgressiveOverrideEnterprise/>

Follow these steps to use the `@override` directive:

1. If the `@override` directive isn't already imported, include it in your schema's `@link` imports:

    ```graphql {3} title="Billing subgraph"
    extend schema

>       @link(url: "https://specs.apollo.dev/federation/v2.3",

Reviewer: ```suggestion
      @link(url: "https://specs.apollo.dev/federation/v2.7",
```

            import: ["@key", "@shareable", "@override"])
    ```

2. Deploy a new version of the Billing subgraph that both defines and resolves the `Bill` fields you want to move:

    <CodeColumns>

      ```graphql title="Payments subgraph"
      type Bill @key(fields: "id") {
        id: ID!
        amount: Int!
        payment: Payment
      }

      type Payment {
        # ...
      }
      ```

      ```graphql {3} title="Billing subgraph"
      type Bill @key(fields: "id") {
        id: ID!
        amount: Int! @override(from: "Payments") #highlight-line
      }
      ```

    </CodeColumns>

      Applying the `@override` directive tells the router to resolve the `amount` field in the Billing subgraph instead of the Payments subgraph.

3. Update your router's supergraph schema to migrate to the updated Billing subgraph.

    * **To migrate gradually with progressive `@override`**, you must use the Apollo Router with an Enterprise license. We recommend using progressive `@override` if it's available.

>        To set progressive `@override` in the Billing subgraph, add a `label` argument to the `@override` directive with the percentage of requests that you want to be resolved by the subgraph. For example:

Reviewer: ```suggestion
       To use progressive `@override`, add a `label` argument to the `@override` directive with the percentage of requests that you want to be resolved by the subgraph. For example:
```

        ```graphql title="Billing subgraph"
        type Bill @key(fields: "id") {
          id: ID!

>           amount: Int! @override(from: "Payments", label:"percent(15)") #highlight-line

Reviewer: ```suggestion
          amount: Int! @override(from: "Payments", label: "percent(15)") #highlight-line
```

        }
        ``` 

>         Setting `label:"percent(15)"` means that 15 percent of the requests for `amount` are resolved by the Billing subgraph, while the remaining 85 percent are resolved by the Payments subgraph. 

Reviewer: ```suggestion
        Setting `label: "percent(15)"` means that 15 percent of the requests for `amount` are resolved by the Billing subgraph, while the remaining 85 percent are resolved by the Payments subgraph. 
```

When the router receives its updated supergraph schema, it immediately starts resolving the `Bill.amount` field from the Billing subgraph while continuing to resolve `Bill.payment` from the Payments subgraph.

<Note>

We can migrate as many entity fields as we want in a single change. To do so, we apply `@override` to every entity field we want to move. We can even migrate entire entities this way.

</Note>

Now that `Bill.amount` is resolved in the Billing subgraph, we can safely _remove_ that field (and its resolver) from the Payments subgraph:

<CodeColumns>

  ```graphql {1-4} title="Payments subgraph"
  type Bill @key(fields: "id") {
    id: ID!
    payment: Payment
  }

  type Payment {
    # ...
  }
  ```

  ```graphql title="Billing subgraph"
  type Bill @key(fields: "id") {
    id: ID!
    amount: Int! @override(from: "Payments")
  }
  ```

</CodeColumns>

After making this change, we deploy our updated Payments subgraph and again update our router's supergraph schema.

<Note>

> Because the router is already _ignoring_ `Bill.amount` in the Payments subgraph thanks to `@override`, we can safely publish our updated schema and deploy the subgraph in any order.

Reviewer: ```suggestion
Because the router is already _ignoring_ `Bill.amount` in the Payments subgraph thanks to `@override`, we can safely publish our updated schema or deploy the subgraph in any order.
```

      ```graphql {3} title="Billing subgraph"
      type Bill @key(fields: "id") {
        id: ID!
        amount: Int! @override(from: "Payments", label: "percent(1)") #highlight-line
      }
      ```

    </CodeColumns>

      Applying the `@override` directive tells the router to resolve the `amount` field in the Billing subgraph instead of the Payments subgraph. 

      Adding a `label` argument to the `@override` directive sets the percentage of traffic to direct to the Billing subgraph. Start with a small percentage. Setting `label: "percent(1)"` means that 1 percent of the requests for `amount` are resolved by the Billing subgraph, while the remaining 99 percent are resolved by the Payments subgraph. 

> 3. Update your router's supergraph schema to migrate to the updated Billing subgraph.

Reviewer: ```suggestion
3. Update your router's supergraph schema to begin the migration to the updated Billing subgraph.
```

> When the router receives its updated supergraph schema, it immediately starts resolving the `Bill.amount` field from the Billing subgraph while continuing to resolve `Bill.payment` from the Payments subgraph.

Reviewer: ```suggestion
When the router receives its updated supergraph schema, it starts resolving the `Bill.amount` field from the Billing subgraph approximately 1% of the time, while continuing to resolve it from the Payments subgraph the other 99%.
```
I changed a couple things here. Worth mentioning I'm dropping the callout to `Bill.payment`, I think it's ok not to mention it here. We can keep it if you think it's helpful though.

3. Update your router's supergraph schema to migrate to the updated Billing subgraph.

   * Use progressive `@override` by adding a `label` argument to the `@override` directive with the percentage of requests that you want to be resolved by the subgraph. For example:

      ```graphql title="Billing subgraph"
      type Bill @key(fields: "id") {
        id: ID!
        amount: Int! @override(from: "Payments", label: "percent(15)") #highlight-line

>       }

Reviewer: The `label` arg should be explained and used initially, so up in step 2 when we first add the override directive. Just for the sake of example, let's use `percent(1)` to illustrate how we should start out small with the migration.

</Note>

> 4. Remove the `@override` directive from the Billing subgraph, because it no longer has any effect:

Reviewer: We need a step that "completes" the migration, i.e. `percent(100)` or remove the label to direct all traffic to the new subgraph. After that happens, then we can remove it from the old one.
